audio-worklet/noisegate-test.html

```diff
@@ -0,0 +1,432 @@
+        let osc = new ConstantSourceNode(context, {offset: 1});
+        osc.connect(gain).connect(noisegate.input);
+        noisegate.output.connect(context.destination);
+
+        // The post-threshold gain should make samples less than the threshold.
+        // It must be greater than 0; otherwise samples will immediately
+        // be 0 after the threshold, and there will be no inspectable delay.
+        const beneathThresholdGain = 0.1;
+        gain.gain.setValueAtTime(0.1, timeToThreshold);
+        
+        osc.start();
+
+        context.startRendering().then((buffer) => {
+          let channel = buffer.getChannelData(0);
+          
+          // With a constant source, the number of steps it takes for the
+          // envelope to open and close is a function of alpha relative
+          // to the threshold.
+          const numOpeningSteps =
+              getNumStepsFromZeroUpTo_(linearThreshold, noisegate.alpha_);
+
+          // If sample rate or alpha changes, add one to numClosingSteps change.
+          const numClosingSteps =
+              getNumStepsFromOneDownTo_(linearThreshold, noisegate.alpha_);
+
+          const openIndex = numOpeningSteps + scriptProcessorLatency;
+          
+          const sourceBeneathThresholdIndex =
+              testFrames * timeToThreshold + scriptProcessorLatency;
+
+          let beforeOpenPhase =
+              channel.slice(scriptProcessorLatency, openIndex);
+          let openPhase = channel.slice(openIndex, sourceBeneathThresholdIndex);
+          let closePhase = channel.slice(
+              sourceBeneathThresholdIndex + numClosingSteps, channel.length);
+          let betweenOpenAndClosePhase =
+              channel.slice(sourceBeneathThresholdIndex,
+                            sourceBeneathThresholdIndex + numClosingSteps);
+
+          should(beforeOpenPhase, 'The samples before the gate opens')
+              .beConstantValueOf(0);
+          should(openPhase, 'The open phase').beConstantValueOf(1);
+          should(closePhase, 'The closed phase').beConstantValueOf(0);
+
+          // The relativeThreshold is set experimentally, and exists because
+          // of the imperfection of floating pointing multiplication. This
+          // test may fail for one sample if settings (e.g. alpha) change.
+          should(betweenOpenAndClosePhase, 'The samples before the gate closes')
+              .beCloseToArray(new Array(betweenOpenAndClosePhase.length)
+                                  .fill(beneathThresholdGain),
+                              {relativeThreshold : 1e-7});
+
+          // With current settings, the opening delay is 0.00049s
+          console.log('The opening delay is ' +
+                      toTime_(numOpeningSteps, sampleRate));
+
+          // With current settings, the closing delay is 0.0034s
+          console.log('The closing delay is ' +
+                      toTime_(numClosingSteps, sampleRate));
```

Comment:
If it's worth printing out these values, it's worth checking they have the expected values. Convert these to test assertions.